In [30]:
from functions import *

# EDA

In [2]:
songs = pd.read_csv('data/spotify_songs.csv')
songs.head(5)


track_id                                         track_name  \
0  0017A6SJgTbfQVU2EtsPNo                                           Pangarap   
1  004s3t0ONYlzxII9PLgU6z                                       I Feel Alive   
2  00chLpzhgVjxs1zKC9UScL                                             Poison   
3  00cqd6ZsSkLZqGMlQCR0Zo  Baby It's Cold Outside (feat. Christina Aguilera)   
4  00emjlCv9azBN0fzuuyLqy                                         Dumb Litty   

      track_artist                                             lyrics  \
0  Barbie's Cradle  Minsan pa Nang ako'y napalingon Hindi ko alam ...   
1    Steady Rollin  The trees, are singing in the wind The sky blu...   
2   Bell Biv DeVoe  NA Yeah, Spyderman and Freeze in full effect U...   
3      CeeLo Green  I really can't stay Baby it's cold outside I'v...   
4             KARD  Get up out of my business You don't keep me fr...   

   track_popularity          track_album_id  \
0                41  1srJQ0njEQgd8w4XSqI4JQ   
1                28  3z04Lb9Dsilqw68SHt6jLB   
2                 0  6oZ6brjB8x3GoeSYdwJdPc   
3                41  3ssspRe42CXkhPxdc12xcp   
4                65  7h5X3xhh3peIK9Y0qI5hbK   

                       track_album_name track_album_release_date  \
0                                  Trip               2001-01-01   
1                           Love & Loss               2017-11-21   
2                                  Gold               2005-01-01   
3                  CeeLo's Magic Moment               2012-10-29   
4  KARD 2nd Digital Single ‘Dumb Litty’               2019-09-22   

                                       playlist_name             playlist_id  \
0                                 Pinoy Classic Rock  37i9dQZF1DWYDQ8wBxd7xt   
1                                  Hard Rock Workout  3YouF0u7waJnolytf9JCXf   
2  Back in the day - R&B, New Jack Swing, Swingbe...  3a9y4eeCJRmG9p4YKfqYIx   
3                                     Christmas Soul  6FZYc2BvF7tColxO8PBShV   
4                                  K-Party Dance Mix  37i9dQZF1DX4RDXswvP6Mj   

   ... loudness mode  speechiness  acousticness  instrumentalness  liveness  \
0  ...  -10.068    1       0.0236       0.27900           0.01170    0.0887   
1  ...   -4.739    1       0.0442       0.01170           0.00994    0.3470   
2  ...   -7.504    0       0.2160       0.00432           0.00723    0.4890   
3  ...   -5.819    0       0.0341       0.68900           0.00000    0.0664   
4  ...   -1.993    1       0.0409       0.03700           0.00000    0.1380   

   valence    tempo  duration_ms  language  
0    0.566   97.091       235440        tl  
1    0.404  135.225       373512        en  
2    0.650  111.904       262467        en  
3    0.405  118.593       243067        en  
4    0.240  130.018       193160        en  

[5 rows x 25 columns]

In [13]:
songs[songs.track_name == 'Poison']

track_id track_name    track_artist  \
2      00chLpzhgVjxs1zKC9UScL     Poison  Bell Biv DeVoe   
237    06yHcjr58IKJnfEK7ko3sD     Poison  Bell Biv DeVoe   
622    0dDifktDnbzmGwliFzI0Ld     Poison  Bell Biv DeVoe   
1705   0RnDu3eYJqbFKz6MHv2ajd     Poison  Bell Biv DeVoe   
3433   1iQpStZpzBb5TOuTDKJKdj     Poison    Alice Cooper   
4878   231mi5BkHyJHxQre25S6oS     Poison  Bell Biv DeVoe   
5419   2donM8sTsqYtRpSAwRdPJQ     Poison  Bell Biv DeVoe   
7429   39J10NL0mFTAdJbapoo2rC     Poison    Alice Cooper   
11776  4ZkVOb0mNnm3MFY2LeyqHY     Poison   Martin Garrix   
13186  5NijSs5dAwaIybq1GaRTIe     Poison    Brent Faiyaz   
13903  5XcZRgJv3zMhTqCyESjQrF     Poison    Alice Cooper   
15464  6m59VvDUi0UQsB2eZ9wVbH     Poison  Bell Biv DeVoe   
16677  73ucpKq91TuejrLHgzDNHK     Poison        Rita Ora   

                                                  lyrics  track_popularity  \
2      NA Yeah, Spyderman and Freeze in full effect U...                 0   
237    NA Yeah, Spyderman and Freeze in full effect U...                36   
622    NA Yeah, Spyderman and Freeze in full effect U...                 0   
1705   NA Yeah, Spyderman and Freeze in full effect U...                48   
3433   Your cruel device Your blood, like ice One loo...                59   
4878   NA Yeah, Spyderman and Freeze in full effect U...                 0   
5419   NA Yeah, Spyderman and Freeze in full effect U...                 1   
7429   Your cruel device Your blood, like ice One loo...                57   
11776                                                NaN                 2   
13186  I feel like there's no such thing as a relatio...                68   
13903  Your cruel device Your blood, like ice One loo...                70   
15464  NA Yeah, Spyderman and Freeze in full effect U...                66   
16677  I could have beer for breakfast My sanity for ...                60   

               track_album_id  \
2      6oZ6brjB8x3GoeSYdwJdPc   
237    3ExqNzkCrQkkknQbrJFqMd   
622    5AR2bbUUDI2XEMarSbs8Cn   
1705   2QC1IsQIUNdEz0zgWanPkN   
3433   0ztezUxR9YMkIP1SFRSnV1   
4878   0PPAtp7bE7Itwl0UtG1mXc   
5419   7r7Vb7w9Fc3kUDNgIlPIxk   
7429   7fP5RXmvSPfAY1EgffEPVx   
11776  2m4U0aYL6QDgbkL5GFbCmn   
13186  17dsllSewkPuKbkGWU4chD   
13903  033cvSPAuSU5ArRfIgQSDU   
15464  6H5mxGUWguDjtQ4Uzd8veD   
16677  6hKHpD9mnIjiJWmAunnQT8   

                                        track_album_name  \
2                                                   Gold   
237                                           RnB 1990's   
622            Bell Biv DeVoe Greatest Hits (Remastered)   
1705   20th Century Masters: The Millennium Collectio...   
3433                              Alice Cooper Classicks   
4878                                   20 #1's: 90's R&B   
5419                                                Gold   
7429                            The Best Of Alice Cooper   
11776                                             Poison   
13186                                             Poison   
13903                                              Trash   
15464                                             Poison   
16677                                             Poison   

      track_album_release_date  \
2                   2005-01-01   
237                 2019-07-19   
622                       2000   
1705                2002-01-01   
3433                      1989   
4878                2015-06-15   
5419                2005-01-01   
7429                2007-05-05   
11776               2015-11-02   
13186               2016-08-03   
13903               1989-07-25   
15464               1990-01-01   
16677               2015-05-18   

                                           playlist_name  \
2      Back in the day - R&B, New Jack Swing, Swingbe...   
237                                  CSR 103:9 (GTA: SA)   
622                                       New Jack Swing   
1705            New Jack Swing - 90s R&B fused w Hip Hop   
3433  

## Observation:

- There seem to be songs that are duplicated in different albums / songs with same name but different artist. 
- Not sure how to fix this


In [14]:
songs.lyrics.isna().value_counts()

False    18194
True       260
Name: lyrics, dtype: int64

In [21]:
songs.dropna(subset=['lyrics'], inplace=True)

In [26]:
songs.lyrics.isna().value_counts()

False    15405
Name: lyrics, dtype: int64

In [25]:
songs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15405 entries, 1 to 18452
Data columns (total 25 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   track_id                  15405 non-null  object 
 1   track_name                15405 non-null  object 
 2   track_artist              15405 non-null  object 
 3   lyrics                    15405 non-null  object 
 4   track_popularity          15405 non-null  int64  
 5   track_album_id            15405 non-null  object 
 6   track_album_name          15405 non-null  object 
 7   track_album_release_date  15405 non-null  object 
 8   playlist_name             15405 non-null  object 
 9   playlist_id               15405 non-null  object 
 10  playlist_genre            15405 non-null  object 
 11  playlist_subgenre         15405 non-null  object 
 12  danceability              15405 non-null  float64
 13  energy                    15405 non-null  float64
 14  key   

## Data Preprocessing

Since we will be using NLP to cluster songs together, we will work on only English songs.

In [27]:
# Filtering dataset to contain only english songs

songs = songs[songs['language'] == 'en']
songs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15405 entries, 1 to 18452
Data columns (total 25 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   track_id                  15405 non-null  object 
 1   track_name                15405 non-null  object 
 2   track_artist              15405 non-null  object 
 3   lyrics                    15405 non-null  object 
 4   track_popularity          15405 non-null  int64  
 5   track_album_id            15405 non-null  object 
 6   track_album_name          15405 non-null  object 
 7   track_album_release_date  15405 non-null  object 
 8   playlist_name             15405 non-null  object 
 9   playlist_id               15405 non-null  object 
 10  playlist_genre            15405 non-null  object 
 11  playlist_subgenre         15405 non-null  object 
 12  danceability              15405 non-null  float64
 13  energy                    15405 non-null  float64
 14  key   

# Observation

We went from 18454 to 15405 songs, it seems like the majority of songs in spotify are in English.

# Cleaning lyrics

Some English songs are in bilingual so we will have to use regular expression to filter out non-latin characters, lowering every characters, removing stopwords.

In [31]:
clean_lyrics(songs.lyrics)

c:\Users\thame\OneDrive\Documents\GitHub\songrecommender\functions.py:13: FutureWarning: The default value of regex will change from True to False in a future version.
  text = text.str.replace('[^a-z]', ' ')


In [32]:
songs.iloc[0]['lyrics']

'trees singing wind sky blue angels smiled saw lonely bench half past four kissed soft soft hands kissed lips angels smiled thought hey feel alive park sign said closed jumped fence cares kissed tree danced midnight sun loved without knowing laughed felt free angels smiled thought hey feel alive'

In [34]:
songs.to_csv('data/en_songs.csv', index=False)